src/main/scala/org/broadinstitute/dsde/agora/server/business/PermissionBusiness.scala

```diff
@@ -1,105 +1,173 @@
-    authorizeNamespaceRequester(getNamespaceACLs(entity, requester), entity, requester)
+  private def authNamespaceRequester[T](db: DataAccess, entity: AgoraEntity, requester: String)(op: => ReadWriteAction[T]): ReadWriteAction[T] = {
+    withNamespaceACLs(db, entity, requester) { namespaceACLs =>
+      authorizeNamespaceRequester(db, namespaceACLs, entity, requester) flatMap { _ => op }
+    }
+
   }
 
-  private def authorizeNamespaceRequester(acls: Seq[AccessControl], entity: AgoraEntity, requester: String): Unit = {
-    if (!acls.exists(_.roles.canManage))
-      throw NamespaceAuthorizationException(AgoraPermissions(Manage), entity, requester)
+  private def authorizeNamespaceRequester(db: DataAccess, acls: Seq[AccessControl], entity: AgoraEntity, requester: String): ReadWriteAction[Unit] = {
+    //NOTE: The use of aePerms here seems wrong, but it's not!
+    db.aePerms.addUserIfNotInDatabase(requester) map { _ =>
+      if (!acls.exists(_.roles.canManage))
+        throw NamespaceAuthorizationException(AgoraPermissions(Manage), entity, requester)
+    }
   }
 
   /**
     * Utility method to both authorize the entity requester and check that the requester is not modifying their own
     * permissions
     */
-  private def authorizeEntityRequester(entity: AgoraEntity, requester: String, accessObject: AccessControl): Unit = {
-    val currentACLs = getEntityACLs(entity, requester)
-    checkSameRequesterAndPermissions(currentACLs.find(acl => acl.user.equals(requester)), accessObject)
-    authorizeEntityRequester(currentACLs, entity, requester)
+  private def authEntityRequester[T](db: DataAccess, entity: AgoraEntity, requester: String, accessObject: AccessControl)(op: => ReadWriteAction[T]): ReadWriteAction[T] = {
+    withEntityACLs(db, entity, requester) { entityACLs =>
+      checkSameRequesterAndPermissions(entityACLs.find(acl => acl.user.equals(requester)), accessObject)
+      authorizeEntityRequester(db, entityACLs, entity, requester) flatMap { _ => op }
+    }
   }
 
-  private def authorizeEntityRequester(entity: AgoraEntity, requester: String): Unit = {
-    authorizeEntityRequester(getEntityACLs(entity, requester), entity, requester)
+  private def authEntityRequester[T](db: DataAccess, entity: AgoraEntity, requester: String)(op: => ReadWriteAction[T]): ReadWriteAction[T] = {
+    withEntityACLs(db, entity, requester) { entityACLs =>
+      authorizeEntityRequester(db, entityACLs, entity, requester) flatMap { _ => op }
+    }
   }
 
-  private def authorizeEntityRequester(acls: Seq[AccessControl], entity: AgoraEntity, requester: String): Unit = {
-    if (!acls.exists(_.roles.canManage))
-      throw AgoraEntityAuthorizationException(AgoraPermissions(Manage), entity, requester)
+  private def authorizeEntityRequester(db: DataAccess, acls: Seq[AccessControl], entity: AgoraEntity, requester: String): ReadWriteAction[Unit] = {
+    db.aePerms.addUserIfNotInDatabase(requester) map { _ =>
```

Comment:
`addUserIfNotInDatabase` used to be called EVERYWHERE in the route helpers. no longer! i moved it to `authNamespaceRequester` and `authEntityRequester` and confirmed that behaviour is correct. so now it happens in the transaction along with everything else. hooray!